caproto/ioc_examples/lakeshore.py

```diff
@@ -0,0 +1,399 @@
+
+    @property
+    def output(self):
+        return self._output
+
+    @property
+    def feedback(self):
+        return self._feedback
+
+    @property
+    def setpoint(self):
+        return self._setpoint
+
+    @setpoint.setter
+    def setpoint(self, value):
+        """
+        Always set ramping to true at the same time as
+        the setpoint change. Is will avoid a race condition,
+        and will allow the client to just check for ramping = False
+        to determine completion.
+        """
+        self._setpoint_target = value
+        self._ramping = True
+        if self.ramp_rate is None:
+            self._setpoint = value
+
+    @property
+    def ramping(self):
+        return self._ramping
+
+    def stop(self):
+        self._run = False
+
+    def _executor(self):
+        iteration_time = 0.1
+        while self._run:
+            self._feedback = self._get_feedback()
+            self._output = self.__call__(self._get_feedback())
+            self.history["output"].append(self._output)
+            self.history["feedback"].append(self._feedback)
+            self.history["setpoint"].append(self.setpoint)
+            self.history["timestamp"].append(time.time())
+            self._set_output(self._output)
+
+            # Ramping logic.
+            remaining = self._setpoint_target - self.setpoint
+            self._ramping = bool(remaining)
+            if isinstance(self.ramp_rate, (int, float)):
+                if remaining > 0:
+                    self._setpoint += min(
+                        self.ramp_rate * iteration_time, abs(remaining)
+                    )
+                elif remaining < 0:
+                    self._setpoint -= min(
+                        self.ramp_rate * iteration_time, abs(remaining)
+                    )
+            elif self.ramp_rate is None and remaining != 0:
+                self._setpoint = self._setpoint_target
+
+            time.sleep(iteration_time)
+
+
+internal_process = contextvars.ContextVar("internal_process", default=False)
+
+
+def no_reentry(func):
+    """
+    This is needed for put completion.
+    """
+
+    @functools.wraps(func)
+    async def inner(*args, **kwargs):
+        if internal_process.get():
+            return
+        try:
+            internal_process.set(True)
+            return await func(*args, **kwargs)
+        finally:
+            internal_process.set(False)
+
+    return inner
+
+
+class LakeshoreIOC(PVGroup):
+    """
+    Simulated Lakeshore IOC with put completion on the setpoint.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._sample = ThermalMaterial()
+
+        self._temperature_controller = PIDController(
+            lambda: self._sample.temperature,
+            self._sample.set_heater_power,
+            Kp=1,
+            Ki=0.1,
+            Kd=0.05,
+            setpoint=150,
+        )
+
+    async def device_poller(self):
+        while True:
```

Comment:
You'll want an `asyncio.sleep()` in here somewhere - it's probably consuming a _lot_ of your CPU while running this...

Reply:
Actually, without the sleep, the rbv don't even work.

Reply:
Added the sleep